File: 算法/代码随想录算法PDF/00调试源代码/39.组合总和.cpp
/*
 * @lc app=leetcode.cn id=39 lang=cpp
 *
 * [39] 组合总和
 */

// @lc code=start
class Solution {
public:
    vector<vector<int>> res;
    vector<int> path;
    void backtracking(vector<int>& candidates, int need, int startIndex) {
        if (need < 0) return ;
        if (need == 0) {
            res.push_back(path);
            return ;
        }
        // 添加判断进行剪枝
        for (int i = startIndex; i < candidates.size() && need - candidates[i] >= 0; i++) {
            path.push_back(candidates[i]);
            backtracking(candidates, need - candidates[i], i);
            path.pop_back();
        }
    }
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        res.clear();
        path.clear();
        sort(candidates.begin(), candidates.end());  // 需要排序
        backtracking(candidates, target, 0);
        return res;
    }
};
// @lc code=end

File: 算法/代码随想录算法PDF/00调试源代码/39.组合总和.py
#
# @lc app=leetcode.cn id=39 lang=python3
#
# [39] 组合总和
#

# @lc code=start
class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        res = []
        path = []
        def backtracking(candidates, need, startIndex):
            nonlocal res, path
            if need < 0: return
            if need == 0:
                res.append(path[:])
                return
            for i in range(startIndex, len(candidates)):
                if need - candidates[i] < 0: return  # 添加判断进行剪枝
                path.append(candidates[i])
                backtracking(candidates, need - candidates[i], i)
                path.pop()
        
        candidates.sort()  # 提前排序
        backtracking(candidates, target, 0)
        return res
# @lc code=end

